onnxruntime/contrib_ops/cuda/transformers/generation_device_helper.cc

```diff
@@ -551,13 +555,31 @@ Status ProcessLogits(const OrtValue& logits,                                 //
   gsl::span<const int32_t> next_tokens(cpu_state->topk_tokens.data(), beam_state->next_tokens.size());
   gsl::span<const int32_t> next_indices(cpu_state->topk_indices.data(), beam_state->next_indices.size());
 
-  // Limitation: beam scorer runs in CPU. It might be better to use CUDA kernel to replace it.
+  // TODO: Implement BeamScorer on CUDA
   beam_scorer->Process(
       sequences,
       next_scores,
       next_tokens,
       next_indices);
 
+  // TODO: This is a temporary work-around as BeamScorer currently only runs on CPU.
+  // We can remove these kinds of work-arounds once BeamScorer runs on CUDA eventually.
+  auto chosen_indices = beam_scorer->GetNextIndices();
+  auto beam_state_chosen_indices = beam_state->chosen_indices;
+
+  if (!beam_state_chosen_indices.empty()) {
+    // If we have allocated `chosen_indices` in beam_state, it means that we
+    // will be needing the chosen indices from BeamScorer as we are using
+    // DecoderMaskedMultiheadAttention, so copy it over.
+    CUDA_RETURN_IF_ERROR(cudaMemcpyAsync(beam_state_chosen_indices.data(),
```

Comment:
Since `BeamScorer` runs on CPU and we need to access to the finalized chosen beam indices for the next iteration to populate the indirection cache, we unfortuantely need to introduce this cross-device copy. Hopefully, once BeamScorer is implemented on CUDA, we can find a way to avoid these cross-device copies. To that effect, left appropriate TODOs in several places in this file.